homeassistant/components/device_tracker/icloud.py

```diff
@@ -47,39 +50,39 @@ def keep_alive(now):
         api.authenticate()
         _LOGGER.info("Authenticate against iCloud")
 
-    track_utc_time_change(hass, keep_alive, second=0)
-
     def update_icloud(now):
         """Authenticate against iCloud and scan for devices."""
         try:
             # The session timeouts if we are not using it so we
             # have to re-authenticate. This will send an email.
-            api.authenticate()
+            keep_alive(None)
             # Loop through every device registered with the iCloud account
             for device in api.devices:
                 status = device.status()
                 location = device.location()
                 # If the device has a location add it. If not do nothing
                 if location:
                     see(
-                        dev_id=re.sub(r"(\s|\W|')",
-                                      '',
-                                      status['name']),
+                        dev_id=slugify(status['name'].replace(' ', '', 99)),
                         host_name=status['name'],
                         gps=(location['latitude'], location['longitude']),
                         battery=status['batteryLevel']*100,
                         gps_accuracy=location['horizontalAccuracy']
                     )
-                else:
-                    # No location found for the device so continue
-                    continue
         except PyiCloudNoDevicesException:
             _LOGGER.info('No iCloud Devices found!')
 
-    track_utc_time_change(
-        hass, update_icloud,
-        minute=range(0, 60, config.get(CONF_INTERVAL, DEFAULT_INTERVAL)),
-        second=0
-    )
+    hass.bus.listen_once(EVENT_HOMEASSISTANT_START, update_icloud)
+
+    interval = config[CONF_INTERVAL]
+    track_utc_time_change(hass, update_icloud, second=0,
+                          minute=range(0, 60, interval))
+
+    # Schedule keepalives between the updates
+    keep_alive_start = KEEPALIVE_INTERVAL
```

Comment:
I am very confused what is going on here. KEEPALIVE_INTERVAL = 4 and interval default = 8.

That means that we are counting from [4…7] till 60 in increments of 8:

4, 12, 20, 28, 36, 44, 52
5, 13, 21, 29, 37, 45, 53
6, 14, 22, 30, 38, 46, 54
7, 15, 23, 31, 39, 47, 55

That seems not a very consistent behavior. There is 9 minutes around hour change and there are gaps of 5 minutes every 4 minutes.

Also, don't forget that each call to `track_utc_time_change` adds another function to be called every second. Better to combine the values and then register once.

I think this should work correctly:

```python
update_minutes = list(range(0, 60, interval))
keep_alive_minutes = (x for x in range(0, 60, KEEPALIVE_INTERVAL) if x not in update_minutes)
```